### PR TITLE
Mirror changes from dotnet/corert

### PR DESCRIPTION
**Pull request by filipnavara**
#22013 + build fixes

### CODE REVIEW

**Comment by filipnavara** [comment]
I've struggled with fixing the build here and in the end took a path of least resistance which involved a change in the shared partition. I had a local branch where I merged `System.Threading.Thread` into `Internal.Runtime.Augments.RuntimeThread` which solves it in a better way, but unfortunately it is not ready to be submitted as PR yet.

File: src/System.Private.CoreLib/src/Internal/Threading/Tasks/AsyncCausalitySupport.cs [hunk 1]
```diff
@@ -57,6 +57,18 @@ public static void TraceOperationCompletedError(Task task)
         {
             AsyncCausalityTracer.TraceOperationCompletion(CausalityTraceLevel.Required, task.Id, AsyncCausalityStatus.Error);
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void TraceSynchronousWorkStart(Task task)
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/Internal/Threading/Tasks/AsyncCausalitySupport.cs]: This type is meant to have just enough async debugger support needed by the AsTask() extension methods in the WindowsRuntimeSystemExtensions class. (See comment at the top.)

I do not think we want to be adding more public methods to it.

Comment (filipnavara) [reply src/System.Private.CoreLib/src/Internal/Threading/Tasks/AsyncCausalitySupport.cs]: Oops, I didn't realize it is public.

Comment (filipnavara) [reply src/System.Private.CoreLib/src/Internal/Threading/Tasks/AsyncCausalitySupport.cs]: I changed the two new methods to be internal. If that's not a good solution I suggest to revert https://github.com/dotnet/corert/pull/6832 and go back to the drawing board. I can probably reshuffle the dependencies in a way that will make it cleaner.

**Comment by filipnavara** [comment]
@dotnet-bot test Windows_NT x64 Checked Innerloop Build and Test

File: src/System.Private.CoreLib/src/Internal/Threading/Tasks/AsyncCausalitySupport.cs (hunk 1 above)
Comment (jkotas) [reply src/System.Private.CoreLib/src/Internal/Threading/Tasks/AsyncCausalitySupport.cs]: This is ok interim solution. The comment at the top of this file is not correct anymore - it should be update accordingly in next iteration if these methods stay.

**Pull request merged**